## Transformação do dataset tarbase_data.csv
- Selecionar apenas as espécies de interesse
- Adicionar uma coluna com as sequencias de cada mirna

In [51]:
# Import
import pandas as pd
import numpy as np

In [110]:
df = pd.read_csv('/content/drive/My Drive/TCC - 2020/tarbase_data.csv', sep='\t')

In [111]:
df.shape

(649378, 12)

In [112]:
df.head()

geneId       geneName  ... up_down condition
0  0910001A06Rik  0910001A06Rik  ...      UP       NaN
1  1200004M23Rik  1200004M23Rik  ...    DOWN       NaN
2  1700027J05Rik  1700027J05Rik  ...      UP       NaN
3  1810015A11Rik  1810015A11Rik  ...    DOWN       NaN
4  2310047A01Rik  2310047A01Rik  ...    DOWN       NaN

[5 rows x 12 columns]

In [113]:
# Filtrando as 5 espécies observadas
df = df.loc[df['species'].isin(['Arabidopsis thaliana','Oryza sativa','Zea mays','Medicago truncatula','Vitis vinifera'])]

In [114]:
df.shape

(2660, 12)

In [115]:
df.head()

geneId            geneName  ... up_down condition
51  AC233070_21.1(mtr)  AC233070_21.1(mtr)  ...    DOWN       NaN
52  AC233070_21.1(mtr)  AC233070_21.1(mtr)  ...    DOWN       NaN
53  AC233100_46.1(mtr)  AC233100_46.1(mtr)  ...    DOWN       NaN
54  AC233680_24.1(mtr)  AC233680_24.1(mtr)  ...    DOWN       NaN
55   AC233685_3.1(mtr)   AC233685_3.1(mtr)  ...    DOWN       NaN

[5 rows x 12 columns]

In [116]:
# Eliminando algumas colunas
df = df.drop(columns=['category','method','positive_negative','direct_indirect','up_down','condition','cell_line','tissue'])
df.head()

geneId            geneName           mirna              species
51  AC233070_21.1(mtr)  AC233070_21.1(mtr)  mtr-miR396a-5p  Medicago truncatula
52  AC233070_21.1(mtr)  AC233070_21.1(mtr)  mtr-miR396b-5p  Medicago truncatula
53  AC233100_46.1(mtr)  AC233100_46.1(mtr)  mtr-miR1507-3p  Medicago truncatula
54  AC233680_24.1(mtr)  AC233680_24.1(mtr)    mtr-miR2643a  Medicago truncatula
55   AC233685_3.1(mtr)   AC233685_3.1(mtr)     mtr-miR159a  Medicago truncatula

In [117]:
df.shape

(2660, 4)

In [118]:
# Verificando quantos miRNA  únicos existem (desconsidera valores repetidos)
df['mirna'].nunique()

316

In [119]:
# Verificando a frequência de cada miRNA
df['mirna'].value_counts()

zma-miR164c-5p         37
zma-miR164d-5p         37
zma-miR169c-5p         37
zma-miR164a-5p         37
zma-miR169r-5p         37
                       ..
osa-miR413              1
osa-miR319a-3p.2-3p     1
ath-miR842              1
zma-miR399i-3p          1
mtr-miR2111i            1
Name: mirna, Length: 316, dtype: int64

## Arquivos fasta (.fa)



In [ ]:
!pip install biopython

In [96]:
from Bio import SeqIO

lista_seq = []
lista_id = []

for fasta in SeqIO.parse('/content/drive/My Drive/TCC - 2020/mature.fa','fasta'):
    
    lista_id.append(fasta.id)
    lista_seq.append(str(fasta.seq))
    
    

    #print(fasta.id)
    #print(fasta.seq)



In [97]:
len(lista_id)

48885

In [98]:
len(lista_seq)

48885

In [121]:
print(lista_seq)

['UGAGGUAGUAGGUUGUAUAGUU', 'CUAUGCAAUUUUCUACCUUACC', 'UCCCUGAGACCUCAAGUGUGA', 'ACACCUGGGCUCUCCGGGUACC', 'CAUACUUCCUUACAUGCCCAUA', 'UGGAAUGUAAAGAAGUAUGUA', 'CAUCAAAGCGGUGGUUGAUGUG', 'UAUCACAGCCAGCUUUGAUGUGC', 'AGGCAGUGUGGUUAGCUGGUUG', 'ACGGCUACCUUCACUGCCACCC', 'UGCUGGUUUCUUCCACAGUGGUA', 'UCACCGGGUGGAAACUAGCAGU', 'CGCCAAUUUUCGCUUCAGUGCUA', 'UCACCGGGUGAAAAUUCGCAUG', 'UGUGGGUGUCCGUUGCGGUGCUA', 'UCACCGGGUGAACACUUGCAGU', 'UCCGGUUUUUUCCGUGGUGAUA', 'UCACCGGGAGAAAAACUGGAGU', 'AGCUGAUUUCGUCUUGGUAAUA', 'UCACCGGGUGUAAAUCAGCUUG', 'AGUGGAUGUAUGCCAUGAUGAUA', 'UCACCGGGUGUACAUCAGCUAA', 'GGUGGUUUUUCUCUGCAGUGAUA', 'UCACCGGGUGAAAAAUCACCUA', 'GUGGGUGUUUGCUUUUUCGGUGAAG', 'UCACCGGGUUAACAUCUACAGA', 'GACAUCAAGAAACUAGUGAUUAUG', 'UAUCACAGUUUACUUGCUGUCGC', 'CUGGAUGUGCUCGUUGGUCAUA', 'UGACUAGAGACACAUUCAGCU', 'CUGGAUGUGCUCGUUAGUCAUA', 'UGACUAGAGACACAUUCAGCU', 'AAGAGAGCCGUCUAUUGACAGU', 'UGUCAUGGAGUCGCUCUCUUCA', 'AAGAGAGCAGUCUAUUGACAGU', 'UGUCAUGGAGGCGCUCUCUUCA', 'UGAGGUAGGCUCAGUAGAUGCGA', 'ACAUCCACCAGCCUAGCUCGCA', 'C

In [102]:
# Criando um dicionário a partir das listas
my_dict = {'id':lista_id,'sequencia':lista_seq}

In [ ]:
print(my_dict)

In [105]:
# Criar um dataframe a partir do dicionário
df_seq = pd.DataFrame(data=my_dict)

In [122]:
df_seq.head()

id               sequencia
0  cel-let-7-5p  UGAGGUAGUAGGUUGUAUAGUU
1  cel-let-7-3p  CUAUGCAAUUUUCUACCUUACC
2  cel-lin-4-5p   UCCCUGAGACCUCAAGUGUGA
3  cel-lin-4-3p  ACACCUGGGCUCUCCGGGUACC
4  cel-miR-1-5p  CAUACUUCCUUACAUGCCCAUA

## Fazendo um merge (inner join) entre o dataframe com os genesId e o dataframe com as sequências

In [123]:
new_df = pd.merge(df, df_seq,left_on='mirna', right_on='id', how='left')

In [124]:
new_df.head()

geneId  ...               sequencia
0  AC233070_21.1(mtr)  ...   UUCCACAGCUUUCUUGAACUU
1  AC233070_21.1(mtr)  ...   UUCCACAGCUUUCUUGAACUG
2  AC233100_46.1(mtr)  ...  CCUCGUUCCAUACAUCAUCUAG
3  AC233680_24.1(mtr)  ...   UUUGGGAUCAGAAAUUAGAGA
4   AC233685_3.1(mtr)  ...   UUUGGAUUGAAGGGAGCUCUA

[5 rows x 6 columns]

In [125]:
# Verificando valores ausentes no dataset
new_df.isnull().sum()

geneId       0
geneName     0
mirna        0
species      0
id           0
sequencia    0
dtype: int64

In [132]:
new_df['sequencia'].describe()

count                     2660
unique                     177
top       UGACAGAAGAGAGUGAGCAC
freq                       252
Name: sequencia, dtype: object

In [133]:
new_df['mirna'].describe()

count               2660
unique               316
top       zma-miR164c-5p
freq                  37
Name: mirna, dtype: object

In [134]:
new_df['id'].describe()

count               2660
unique               316
top       zma-miR164c-5p
freq                  37
Name: id, dtype: object

**Podemos observar a partir dos valores unicos descritos acima que existem 316 mirna's unicos, porém só 177 sequências unicas. Logo, 139 sequências estão associadas a mais de um mirna.**

In [149]:
# Verificando a quantidade de sequencia em cada mirna
pd.set_option("display.max_rows", 400)
new_df.groupby(['species','mirna'])['sequencia'].count()

species               mirna              
Arabidopsis thaliana  ath-miR163              5
                      ath-miR170-3p           1
                      ath-miR173-5p           5
                      ath-miR400              4
                      ath-miR403-3p           1
                      ath-miR408-3p           2
                      ath-miR472-3p           2
                      ath-miR774a             1
                      ath-miR775              1
                      ath-miR778              1
                      ath-miR781a             1
                      ath-miR822-5p           2
                      ath-miR823              2
                      ath-miR824-5p           1
                      ath-miR826a             1
                      ath-miR827              1
                      ath-miR837-3p           1
                      ath-miR838              1
                      ath-miR842              1
                      ath-miR844-5p           

In [141]:
new_df.describe()

geneId       geneName  ...              id             sequencia
count            2660           2660  ...            2660                  2660
unique            417            417  ...             316                   177
top     GRMZM2G165488  GRMZM2G165488  ...  zma-miR164c-5p  UGACAGAAGAGAGUGAGCAC
freq               68             68  ...              37                   252

[4 rows x 6 columns]

In [150]:
new_df.loc[new_df['mirna'] == 'ath-miR173-5p']

geneId   geneName  ...             id               sequencia
12  At1g50055  At1g50055  ...  ath-miR173-5p  UUCGCUUGCAGAGAGAAAUCAC
28  AT2G27400  AT2G27400  ...  ath-miR173-5p  UUCGCUUGCAGAGAGAAAUCAC
30  At2g39675  At2g39675  ...  ath-miR173-5p  UUCGCUUGCAGAGAGAAAUCAC
31  AT2G39681  AT2G39681  ...  ath-miR173-5p  UUCGCUUGCAGAGAGAAAUCAC
45  AT5G10950  AT5G10950  ...  ath-miR173-5p  UUCGCUUGCAGAGAGAAAUCAC

[5 rows x 6 columns]

In [152]:
# Verificando a quantidade de especies em cada sequencia
pd.set_option("display.max_rows", 400)
new_df.groupby(['sequencia','mirna'])['species'].count()

sequencia                 mirna              
AAACGAACAAAAAACUGAUGG     ath-miR837-3p           1
AAAGUUAGAGAAGUUUGACUU     osa-miR437              1
AAAGUUCAGUCUUCAUAGUAUC    mtr-miR2669a            1
AACAGAACAGAAGCAGAGCAG     osa-miR415              2
AAGAUAAGCGCCUUAGUUCUG     ath-miR852              1
AAGCCAAGGAUGACUUGCCGG     mtr-miR169d-5p          2
AAGCCAAGGAUGACUUGCCUA     mtr-miR169f             2
AAGCUCAGGAGGGAUAGCGCC     osa-miR390-5p           2
                          vvi-miR390              1
                          zma-miR390a-5p          4
                          zma-miR390b-5p          4
ACAUUAUGGGACGGAGGGAGU     osa-miR1436            10
AGAAGAGAGAGAGUACAGCCU     zma-miR529-5p          26
AGAAUCCUGAUGAUGCUGCAG     mtr-miR172a             4
AGAAUCUUGAUGAUGAUUAAA     osa-miR1426             1
AGAAUCUUGAUGAUGCUGCA      zma-miR172a            33
                          zma-miR172b-3p         33
                          zma-miR172c-3p         33
                  